extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcAuthenticationMechanism.java

```diff
@@ -0,0 +1,43 @@
+package io.quarkus.oidc.runtime;
+
+import java.util.concurrent.CompletionStage;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+import io.quarkus.security.identity.IdentityProviderManager;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.vertx.http.runtime.security.ChallengeData;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticationMechanism;
+import io.vertx.ext.web.RoutingContext;
+
+@ApplicationScoped
+public class OidcAuthenticationMechanism implements HttpAuthenticationMechanism {
+
+    @Inject
+    DefaultTenantConfigResolver tenantConfigResolver;
+    private BearerAuthenticationMechanism bearerAuth = new BearerAuthenticationMechanism();
+    private CodeAuthenticationMechanism codeAuth = new CodeAuthenticationMechanism();
+
+    @Override
+    public CompletionStage<SecurityIdentity> authenticate(RoutingContext context,
+            IdentityProviderManager identityProviderManager) {
+        TenantConfigContext configContext = tenantConfigResolver.resolve(context);
```

Comment:
@pedroigor, @stuartwdouglas  Hi, this is a work in progress, I have one more test in the works which combines the service and web-app tenants. Happy enough with how it is working out, the only problem is resolving the configContext here deadlocks in `oidc-tenancy` tests :-) in the sync lock `DefaultTenantConfigResolver`.
Pretty sure it can be reproduced if for example `BearerAuthenticationMechanism.authenticate` on the master  would try to resolve early in its `authenticate`. Nearly there though...Thanks

Reply:
I've been wondering today how I'm going to resolve this deadlock  and realized what is really needed here is a resolution without attempting to create `OAuth2Auth`, at this point only a check what kind of application it is is needed. `DefaultTenantConfigResolver.resolve` needs a `boolean create` parameter, same one which is used by `isBlocking` check internally :-)

Reply:
OK. So only tests are missing. Otherwise, LGTM. Thanks.